src/util/ExpressionUtils.cpp

```diff
@@ -76,18 +76,125 @@ std::unique_ptr<InputPropertyExpression> ExpressionUtils::inputPropExpr(const st
 }
 
 std::unique_ptr<Expression>
-ExpressionUtils::pushOrs(const std::vector<std::unique_ptr<RelationalExpression>>& rels) {
+ExpressionUtils::pushOrs(const std::vector<std::unique_ptr<Expression>>& rels) {
+    return pushImpl(Expression::Kind::kLogicalOr, rels);
+}
+
+std::unique_ptr<Expression>
+ExpressionUtils::pushAnds(const std::vector<std::unique_ptr<Expression>>& rels) {
+    return pushImpl(Expression::Kind::kLogicalAnd, rels);
+}
+
+std::unique_ptr<Expression>
+ExpressionUtils::pushImpl(Expression::Kind kind,
+                          const std::vector<std::unique_ptr<Expression>>& rels) {
     DCHECK_GT(rels.size(), 1);
-    auto root = std::make_unique<LogicalExpression>(Expression::Kind::kLogicalOr);
+    DCHECK(kind == Expression::Kind::kLogicalOr || kind == Expression::Kind::kLogicalAnd);
+    auto root = std::make_unique<LogicalExpression>(kind);
     root->addOperand(rels[0]->clone().release());
     root->addOperand(rels[1]->clone().release());
     for (size_t i = 2; i < rels.size(); i++) {
-        auto l = std::make_unique<LogicalExpression>(Expression::Kind::kLogicalOr);
+        auto l = std::make_unique<LogicalExpression>(kind);
         l->addOperand(root->clone().release());
         l->addOperand(rels[i]->clone().release());
         root = std::move(l);
     }
     return root;
 }
+
+std::unique_ptr<Expression> ExpressionUtils::expandExpr(const Expression *expr) {
+    auto kind = expr->kind();
+    std::vector<std::unique_ptr<Expression>> target;
+    switch (kind) {
+        case Expression::Kind::kLogicalOr: {
+            const auto *logic = static_cast<const LogicalExpression*>(expr);
+            for (const auto& e : logic->operands()) {
+                if (e->kind() == Expression::Kind::kLogicalAnd) {
+                    target.emplace_back(expandImplAnd(e.get()));
+                } else {
+                    target.emplace_back(expandExpr(e.get()));
+                }
+            }
+            break;
+        }
+        case Expression::Kind::kLogicalAnd: {
+            target.emplace_back(expandImplAnd(expr));
+            break;
+        }
+        default: {
+            return expr->clone();
+        }
+    }
+    if (target.size() == 1) {
+        if (target[0]->kind() == Expression::Kind::kLogicalAnd) {
+            const auto *logic = static_cast<const LogicalExpression*>(target[0].get());
+            const auto& ops = logic->operands();
+            DCHECK_EQ(ops.size(), 2);
+            if (ops[0]->kind() == Expression::Kind::kLogicalOr ||
+                ops[1]->kind() == Expression::Kind::kLogicalOr) {
+                return expandExpr(target[0].get());
```

Comment:
Would you give an example when do we hit this path? I suppose all expression of op has been expanded in `expandImplAnd`.

Reply:
> Would you give an example when do we hit this path? I suppose all expression of op has been expanded in `expandImplAnd`.

t1.c1 == 1 and t1.c2 == 2 and (t1.c4 == 4 or t1.c5 == 5)

```
nebula::graph::ExpressionUtils::expandExpr ExpressionUtils.cpp:135
nebula::graph::ExpressionUtils::expandImplAnd ExpressionUtils.cpp:156
nebula::graph::ExpressionUtils::expandExpr ExpressionUtils.cpp:121
nebula::graph::ExpressionUtilsTest_expandExpression_Test::TestBody ExpressionUtilsTest.cpp:507
void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) 0x0000000002f58ce3
void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) 0x0000000002f5340c
testing::Test::Run() 0x0000000002f38e9c
testing::TestInfo::Run() 0x0000000002f39718
testing::TestCase::Run() 0x0000000002f39d69
testing::internal::UnitTestImpl::RunAllTests() 0x0000000002f4089a
bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) 0x0000000002f59dea
bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) 0x0000000002f541f0
testing::UnitTest::Run() 0x0000000002f3f5a0
RUN_ALL_TESTS() 0x0000000002f65fdc
main 0x0000000002f65f6a
__libc_start_main 0x00007ffff7c86f43
_start 0x0000000001e0142e
```

Reply:
Got it. Convert `AB(C + D)` to `ABC + ABD)`